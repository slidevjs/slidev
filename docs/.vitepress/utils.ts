import { data as features } from '../features/index.data.js'
import { Advanced, Guides } from './pages'

function removeHash(link: string) {
  const idx = link.lastIndexOf('#')
  return idx < 0 ? link : link.slice(0, idx)
}

function getGuideTitle(id: string) {
  return Guides.find(g => g.link.endsWith(`/${id}`))?.text ?? Advanced.find(g => g.link.endsWith(id))?.text ?? id
}

export function resolveLink(link: string): {
  kind: 'external' | 'features' | 'guide'
  url: string
  title?: string
  tags?: string[]
  descripton?: string
} {
  const [kind, nameWithHash] = link.split('/')
  const name = removeHash(nameWithHash)
  switch (kind) {
    case 'http:':
    case 'https:':
    case 'mailto:':
      return { kind: 'external', url: link }
    case 'features': {
      const feature = features[name]
      if (!feature)
        throw new Error(`Feature "${name}" not found.`)
      return {
        kind: 'features',
        title: `âœ¨ ${feature.title}`,
        tags: feature.tags,
        descripton: feature.description,
        url: `/features/${nameWithHash}`,
      }
    }
    case 'guide': {
      return {
        kind: 'guide',
        title: `ðŸ“–  ${getGuideTitle(name)}`,
        tags: ['guide'],
        descripton: 'Click to read this guide',
        url: `/guide/${nameWithHash}`,
      }
    }
    default:
      throw new Error(`Invalid link: ${link}`)
  }
}
