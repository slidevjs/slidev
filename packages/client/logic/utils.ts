import { useTimestamp } from '@vueuse/core'
import { computed, ref } from 'vue'

export function useTimer() {
  const tsStart = ref(Date.now())
  const now = useTimestamp({
    interval: 1000,
  })
  const timer = computed(() => {
    const passed = (now.value - tsStart.value) / 1000
    const sec = Math.floor(passed % 60).toString().padStart(2, '0')
    const min = Math.floor(passed / 60).toString().padStart(2, '0')
    return `${min}:${sec}`
  })
  function resetTimer() {
    tsStart.value = now.value
  }

  return {
    timer,
    resetTimer,
  }
}

export function makeId(length = 5) {
  const result = []
  const characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789'
  const charactersLength = characters.length
  for (let i = 0; i < length; i++)
    result.push(characters.charAt(Math.floor(Math.random() * charactersLength)))
  return result.join('')
}
