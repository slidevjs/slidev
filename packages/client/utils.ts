import type { RouteRecordRaw } from 'vue-router'

export function getSlideClass(route?: RouteRecordRaw, extra = '') {
  const classes = ['slidev-page', extra]

  const no = route?.meta?.slide?.no
  if (no != null)
    classes.push(`slidev-page-${no}`)

  return classes.filter(Boolean).join(' ')
}
