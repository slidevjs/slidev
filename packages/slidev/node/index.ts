import './declare'

export * from './server'
export * from './plugins/preset'
export * from './options'
export { parser } from './parser'
