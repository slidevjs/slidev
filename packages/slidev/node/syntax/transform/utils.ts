export function normalizeRangeStr(rangeStr = '') {
  return !rangeStr.trim() ? [] : rangeStr.trim().split(/\|/g).map(i => i.trim())
}

/**
 * Escape `{{` in code block to prevent Vue interpret it, #99, #1316
 */
export function escapeVueInCode(md: string) {
  return md.replace(/{{/g, '&lbrace;&lbrace;')
}
