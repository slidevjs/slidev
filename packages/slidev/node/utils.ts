import type Token from 'markdown-it/lib/token'
import type { ResolvedFontOptions } from '@slidev/types'
import { satisfies } from 'semver'
import { version } from '../package.json'

export function stringifyMarkdownTokens(tokens: Token[]) {
  return tokens.map(token => token.children
    ?.filter(t => ['text', 'code_inline'].includes(t.type) && !t.content.match(/^\s*$/))
    .map(t => t.content.trim())
    .join(' '))
    .filter(Boolean)
    .join(' ')
}

export function generateGoogleFontsUrl(options: ResolvedFontOptions) {
  const weights = options.weights
    .flatMap(i => options.italic ? [`0,${i}`, `1,${i}`] : [`${i}`])
    .sort()
    .join(';')
  const fonts = options.webfonts
    .map(i => `family=${i.replace(/^(['"])(.*)\1$/, '$1').replace(/\s+/g, '+')}:${options.italic ? 'ital,' : ''}wght@${weights}`)
    .join('&')

  return `https://fonts.googleapis.com/css2?${fonts}&display=swap`
}

export function checkEngine(name: string, engines: { slidev?: string } = {}) {
  if (engines.slidev && !satisfies(version, engines.slidev, { includePrerelease: true }))
    throw new Error(`[slidev] addon "${name}" requires Slidev version range "${engines.slidev}" but found "${version}"`)
}

export function makeId(length = 5) {
  const result = []
  const characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789'
  const charactersLength = characters.length
  for (let i = 0; i < length; i++)
    result.push(characters.charAt(Math.floor(Math.random() * charactersLength)))
  return result.join('')
}
