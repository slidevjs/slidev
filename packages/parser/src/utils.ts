import { isNumber, range, uniq } from '@antfu/utils'

/**
 * 1,3-5,8 => [1, 3, 4, 5, 8]
 */
export function parseRangeString(total: number, rangeStr?: string) {
  if (!rangeStr || rangeStr === 'all' || rangeStr === '*')
    return range(1, total + 1)

  const pages: number[] = []
  for (const part of rangeStr.split(/[,;]/g)) {
    if (!part.includes('-')) {
      pages.push(+part)
    }
    else {
      const [start, end] = part.split('-', 2)
      pages.push(
        ...range(+start, !end ? (total + 1) : (+end + 1)),
      )
    }
  }

  return uniq(pages).filter(i => i <= total).sort((a, b) => a - b)
}

/**
 * Accepts `16/9` `1:1` `3x4`
 */
export function parseAspectRatio(str: string | number) {
  if (isNumber(str))
    return str
  if (!Number.isNaN(+str))
    return +str
  const [wStr = '', hStr = ''] = str.split(/[:\/x\|]/)
  const w = Number.parseFloat(wStr.trim())
  const h = Number.parseFloat(hStr.trim())

  if (Number.isNaN(w) || Number.isNaN(h) || h === 0)
    throw new Error(`Invalid aspect ratio "${str}"`)

  return w / h
}
