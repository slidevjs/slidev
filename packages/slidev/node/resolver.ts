import { dirname, join, resolve } from 'node:path'
import * as fs from 'node:fs'
import process from 'node:process'
import { fileURLToPath } from 'node:url'
import { ensurePrefix, slash } from '@antfu/utils'
import isInstalledGlobally from 'is-installed-globally'
import { resolveGlobal } from 'resolve-global'
import { findClosestPkgJsonPath, findDepPkgJsonPath } from 'vitefu'
import { resolvePath } from 'mlly'
import globalDirs from 'global-directory'
import prompts from 'prompts'
import { parseNi, run } from '@antfu/ni'
import { underline, yellow } from 'kolorist'

export function toAtFS(path: string) {
  return `/@fs${ensurePrefix('/', slash(path))}`
}

/**
 * Find the actual path of the import. If Slidev is installed globally, it will also search globally.
 */
export async function resolveImportPath(importName: string, ensure: true): Promise<string>
export async function resolveImportPath(importName: string, ensure?: boolean): Promise<string | undefined>
export async function resolveImportPath(importName: string, ensure = false) {
  try {
    return await resolvePath(importName, {
      url: import.meta.url,
    })
  }
  catch { }

  if (isInstalledGlobally) {
    try {
      return resolveGlobal(importName)
    }
    catch { }
  }

  if (ensure)
    throw new Error(`Failed to resolve package "${importName}"`)
}

/**
 * Find the root of the package. If Slidev is installed globally, it will also search globally.
 */
export async function findPkgRoot(dep: string, parent: string, ensure: true): Promise<string>
export async function findPkgRoot(dep: string, parent: string, ensure?: boolean): Promise<string | undefined>
export async function findPkgRoot(dep: string, parent: string, ensure = false) {
  const pkgJsonPath = await findDepPkgJsonPath(dep, parent)
  const path = pkgJsonPath ? dirname(pkgJsonPath) : isInstalledGlobally ? findGlobalPkgRoot(dep, false) : undefined
  if (ensure && !path)
    throw new Error(`Failed to resolve package "${dep}"`)
  return path
}

export async function findGlobalPkgRoot(name: string, ensure: true): Promise<string>
export async function findGlobalPkgRoot(name: string, ensure?: boolean): Promise<string | undefined>
export async function findGlobalPkgRoot(name: string, ensure = false) {
  const yarnPath = join(globalDirs.yarn.packages, name)
  if (fs.existsSync(`${yarnPath}/package.json`))
    return yarnPath
  const npmPath = join(globalDirs.npm.packages, name)
  if (fs.existsSync(`${npmPath}/package.json`))
    return npmPath
  if (ensure)
    throw new Error(`Failed to resolve global package "${name}"`)
}

export async function resolveEntry(entryRaw: string, roots: RootsInfo) {
  if (!fs.existsSync(entryRaw) && !entryRaw.endsWith('.md') && !/\/\\/.test(entryRaw))
    entryRaw += '.md'
  const entry = entryRaw.startsWith('@/')
    ? join(roots.userRoot, entryRaw.slice(2))
    : resolve(process.cwd(), entryRaw)
  if (!fs.existsSync(entry)) {
    const { create } = await prompts({
      name: 'create',
      type: 'confirm',
      initial: 'Y',
      message: `Entry file ${yellow(`"${entry}"`)} does not exist, do you want to create it?`,
    })
    if (create)
      fs.copyFileSync(resolve(roots.cliRoot, 'template.md'), entry)
    else
      process.exit(0)
  }
  return entry
}

/**
 * Create a resolver for theme or addon
 */
export function createResolver(type: 'theme' | 'addon', officials: Record<string, string>) {
  async function promptForInstallation(pkgName: string) {
    const { confirm } = await prompts({
      name: 'confirm',
      initial: 'Y',
      type: 'confirm',
      message: `The ${type} "${pkgName}" was not found ${underline(isInstalledGlobally ? 'globally' : 'in your project')}, do you want to install it now?`,
    })

    if (!confirm)
      process.exit(1)

    if (isInstalledGlobally)
      await run(parseNi, ['-g', pkgName])
    else
      await run(parseNi, [pkgName])
  }

  return async function (name: string, importer: string): Promise<[name: string, root: string | null]> {
    const { userRoot } = await getRoots()

    if (name === 'none')
      return ['', null]

    // local path
    if (name[0] === '/')
      return [name, name]
    if (name.startsWith('@/'))
      return [name, join(userRoot, name.slice(2))]
    if (name[0] === '.' || (name[0] !== '@' && name.includes('/')))
      return [name, join(dirname(importer), name)]

    // definitely a package name
    if (name.startsWith(`@slidev/${type}-`) || name.startsWith(`slidev-${type}-`)) {
      const pkgRoot = await findPkgRoot(name, importer)
      if (!pkgRoot)
        await promptForInstallation(name)
      return [name, await findPkgRoot(name, importer, true)]
    }

    // search for local packages first
    {
      const possiblePkgNames = [
        `@slidev/${type}-${name}`,
        `slidev-${type}-${name}`,
        name,
      ]

      for (const pkgName of possiblePkgNames) {
        const pkgRoot = await findPkgRoot(pkgName, importer)
        if (pkgRoot)
          return [pkgName, pkgRoot]
      }
    }

    // fallback to prompt install
    const pkgName = officials[name] ?? (name[0] === '@' ? name : `slidev-${type}-${name}`)
    await promptForInstallation(pkgName)
    return [pkgName, await findPkgRoot(pkgName, importer, true)]
  }
}

async function getUserRoot() {
  const pkgJsonPath = await findClosestPkgJsonPath(process.cwd())
  return pkgJsonPath ? dirname(pkgJsonPath) : process.cwd()
}

function getUserPkgJson(userRoot: string) {
  const path = resolve(userRoot, 'package.json')
  if (fs.existsSync(path))
    return JSON.parse(fs.readFileSync(path, 'utf-8')) as Record<string, any>
  return {}
}

// npm: https://docs.npmjs.com/cli/v7/using-npm/workspaces#installing-workspaces
// yarn: https://classic.yarnpkg.com/en/docs/workspaces/#toc-how-to-use-it
function hasWorkspacePackageJSON(root: string): boolean {
  const path = join(root, 'package.json')
  try {
    fs.accessSync(path, fs.constants.R_OK)
  }
  catch {
    return false
  }
  const content = JSON.parse(fs.readFileSync(path, 'utf-8')) || {}
  return !!content.workspaces
}

function hasRootFile(root: string): boolean {
  // https://github.com/vitejs/vite/issues/2820#issuecomment-812495079
  const ROOT_FILES = [
    // '.git',

    // https://pnpm.js.org/workspaces/
    'pnpm-workspace.yaml',

    // https://rushjs.io/pages/advanced/config_files/
    // 'rush.json',

    // https://nx.dev/latest/react/getting-started/nx-setup
    // 'workspace.json',
    // 'nx.json'
  ]

  return ROOT_FILES.some(file => fs.existsSync(join(root, file)))
}

/**
 * Search up for the nearest workspace root
 */
function searchForWorkspaceRoot(
  current: string,
  root = current,
): string {
  if (hasRootFile(current))
    return current
  if (hasWorkspacePackageJSON(current))
    return current

  const dir = dirname(current)
  // reach the fs root
  if (!dir || dir === current)
    return root

  return searchForWorkspaceRoot(dir, root)
}

export interface RootsInfo {
  cliRoot: string
  clientRoot: string
  userRoot: string
  userPkgJson: Record<string, any>
  userWorkspaceRoot: string
}

let rootsInfo: RootsInfo | null = null

export async function getRoots(): Promise<RootsInfo> {
  if (rootsInfo)
    return rootsInfo
  const cliRoot = fileURLToPath(new URL('..', import.meta.url))
  const clientRoot = await findPkgRoot('@slidev/client', cliRoot, true)
  const userRoot = await getUserRoot()
  const userPkgJson = getUserPkgJson(userRoot)
  const userWorkspaceRoot = searchForWorkspaceRoot(userRoot)
  rootsInfo = {
    cliRoot,
    clientRoot,
    userRoot,
    userPkgJson,
    userWorkspaceRoot,
  }
  return rootsInfo
}
