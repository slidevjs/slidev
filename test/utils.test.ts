import type { ResolvedFontOptions, SlideInfo } from '@slidev/types'
import { relative, resolve } from 'node:path'
import { slash } from '@antfu/utils'
import MarkdownIt from 'markdown-it'
import { describe, expect, it } from 'vitest'
import YAML from 'yaml'
import { parseAspectRatio, parseRangeString } from '../packages/parser/src'
import { getRoots } from '../packages/slidev/node/resolver'
import { generateGoogleFontsUrl, stringifyMarkdownTokens, updateFrontmatterPatch } from '../packages/slidev/node/utils'

describe('utils', () => {
  it('page-range', () => {
    expect(parseRangeString(5)).toEqual([1, 2, 3, 4, 5])
    expect(parseRangeString(5, 'all')).toEqual([1, 2, 3, 4, 5])
    expect(parseRangeString(2, '*')).toEqual([1, 2])
    expect(parseRangeString(5, '1')).toEqual([1])
    expect(parseRangeString(10, '1,2-3,5')).toEqual([1, 2, 3, 5])
    expect(parseRangeString(10, '1;2-3;5')).toEqual([1, 2, 3, 5])
    expect(parseRangeString(10, '6-')).toEqual([6, 7, 8, 9, 10])
  })

  it('aspect-ratio', () => {
    expect(parseAspectRatio(1)).toEqual(1)
    expect(parseAspectRatio('16/9')).toEqual(16 / 9)
    expect(parseAspectRatio('16 / 9 ')).toEqual(16 / 9)
    expect(parseAspectRatio('3:4')).toEqual(3 / 4)
    expect(parseAspectRatio('1x1')).toEqual(1)
    expect(parseAspectRatio('1')).toEqual(1)

    expect(() => parseAspectRatio('hello')).toThrow()
    expect(() => parseAspectRatio('1/0')).toThrow()
  })

  it('stringify-markdown-tokens', () => {
    const md = MarkdownIt({ html: true })
    const stringify = (src: string) => stringifyMarkdownTokens(md.parseInline(src, {}))

    expect(stringify('<span style="color:red">Composable</span> Vue')).toBe('Composable Vue')
    expect(stringify('<b>Whatever</b>')).toBe('Whatever')
    expect(stringify('Talk about `<details/>`')).toBe('Talk about <details/>')
    expect(stringify('What is Readonly\\<T\\> in TypeScript')).toBe('What is Readonly<T> in TypeScript')
    expect(stringify('Welcome to<br />*Slidev*')).toBe('Welcome to Slidev')
  })

  it('google-fonts', () => {
    expect(
      generateGoogleFontsUrl({
        webfonts: ['Fira Code', 'PT Serif'],
        weights: ['200', '400', '600'],
        provider: 'google',
      } as ResolvedFontOptions),
    ).toMatchSnapshot()

    expect(
      generateGoogleFontsUrl({
        webfonts: ['Fira Code', 'PT Serif'],
        weights: ['200', '400', '600'],
        italic: true,
        provider: 'google',
      } as ResolvedFontOptions),
    ).toMatchSnapshot()
  })

  it('roots', async () => {
    const { cliRoot, clientRoot, userRoot, userWorkspaceRoot } = await getRoots(resolve('slides.md'))
    const expectRelative = (v: string) => expect(slash(relative(__dirname, v)))
    expectRelative(cliRoot).toMatchInlineSnapshot(`"../packages/slidev"`)
    expectRelative(clientRoot).toMatchInlineSnapshot(`"../packages/client"`)
    expectRelative(userRoot).toMatchInlineSnapshot(`".."`)
    expectRelative(userWorkspaceRoot).toMatchInlineSnapshot(`".."`)
  })

  it('update frontmatter patch', async () => {
    const dragPos = {
      foo: '1,2,3,4',
    }
    function createFakeSource(yaml: string) {
      const doc = YAML.parseDocument(yaml)
      return {
        frontmatter: {},
        source: {
          frontmatter: doc.toJSON() || {},
          frontmatterRaw: yaml,
          frontmatterDoc: doc,
        },
      } as SlideInfo
    }
    function expectFrontmatter(slide: SlideInfo) {
      return expect(slide.source.frontmatterDoc?.toString())
    }

    const slide1 = createFakeSource(``)
    updateFrontmatterPatch(slide1, { dragPos })
    expectFrontmatter(slide1).toMatchInlineSnapshot(`
      "dragPos:
        foo: 1,2,3,4
      "
    `)

    const slide2 = createFakeSource(`
      # comment
      title: Hello  # another comment
      dragPos:
        bar: 5,6,7,8
    `)
    updateFrontmatterPatch(slide2, { dragPos })
    expectFrontmatter(slide2).toMatchInlineSnapshot(`
      "# comment
      title: Hello # another comment
      dragPos:
        foo: 1,2,3,4
      "
    `)

    // remove a field
    const slide3 = createFakeSource(`
      # comment
      title: Hello  # another comment
      dragPos:
        bar: 5,6,7,8
    `)
    updateFrontmatterPatch(slide3, { title: null })
    expectFrontmatter(slide3).toMatchInlineSnapshot(`
      "dragPos:
        bar: 5,6,7,8
      "
    `)
  })
})
