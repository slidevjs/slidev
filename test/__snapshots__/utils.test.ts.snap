// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`utils > google-fonts 1`] = `"https://fonts.googleapis.com/css2?family=Fira+Code:wght@200;400;600&family=PT+Serif:wght@200;400;600&display=swap"`;

exports[`utils > google-fonts 2`] = `"https://fonts.googleapis.com/css2?family=Fira+Code:ital,wght@0,200;0,400;0,600;1,200;1,400;1,600&family=PT+Serif:ital,wght@0,200;0,400;0,600;1,200;1,400;1,600&display=swap"`;
